jt graph --describe

chrisseaton opened the pull request:
Useful for getting a quick idea of a graph when testing.

```
chrisseaton@Shopifys-MacBook-Pro truffleruby % jt graph --describe linear.rb  
Using TruffleRuby with Graal: mxbuild/truffleruby-jvm-ce
Dumping IGV graphs in /Users/chrisseaton/Documents/graalvm/truffleruby/graal_dumps/2021.11.23.18.58.26.623
[engine] opt done   id=1395  Object#foo                                         |Tier 2|Time   304( 132+172 )ms|AST   17|Inlined   0Y   0N|IR     36/    84|CodeSize     379|Timestamp 22947502019610|Src linear.rb:1
$ /Users/chrisseaton/Documents/graalvm/truffleruby/tool/describe-graph 'graal_dumps/2021.11.23.18.58.26.623/TruffleHotSpotCompilation-12376[Object#foo].bgv' 4
36 nodes, linear

chrisseaton@Shopifys-MacBook-Pro truffleruby % jt graph --describe branches.rb 
Using TruffleRuby with Graal: mxbuild/truffleruby-jvm-ce
Dumping IGV graphs in /Users/chrisseaton/Documents/graalvm/truffleruby/graal_dumps/2021.11.23.18.58.35.073
[engine] opt done   id=1395  Object#foo                                         |Tier 2|Time   371( 165+205 )ms|AST   29|Inlined   0Y   0N|IR     67/   146|CodeSize     542|Timestamp 22955987879827|Src branches.rb:1
[engine] opt deopt  id=1395  Object#foo                                                                                                                                     |Timestamp 22956039884025|Src branches.rb:1
$ /Users/chrisseaton/Documents/graalvm/truffleruby/tool/describe-graph 'graal_dumps/2021.11.23.18.58.35.073/TruffleHotSpotCompilation-12603[Object#foo].bgv' 4
67 nodes, branches

chrisseaton@Shopifys-MacBook-Pro truffleruby % jt graph --describe loops.rb   
Using TruffleRuby with Graal: mxbuild/truffleruby-jvm-ce
Dumping IGV graphs in /Users/chrisseaton/Documents/graalvm/truffleruby/graal_dumps/2021.11.23.18.58.45.760
[engine] opt done   id=1399  block in Object#foo <split-1399>                   |Tier 2|Time   405( 164+241 )ms|AST   11|Inlined   0Y   0N|IR     77/   166|CodeSize     533|Timestamp 22966711043625|Src loops.rb:3
[engine] opt done   id=1396  Object#foo                                         |Tier 2|Time   326( 186+140 )ms|AST   29|Inlined   2Y   0N|IR    140/   215|CodeSize     614|Timestamp 22967120625377|Src loops.rb:1
[engine] opt deopt  id=1398  Integer#times                                                                                                                                  |Timestamp 22967164304590|Src integer.rb:143
[engine] opt deopt  id=1396  Object#foo                                                                                                                                     |Timestamp 22967166243313|Src loops.rb:1
$ /Users/chrisseaton/Documents/graalvm/truffleruby/tool/describe-graph 'graal_dumps/2021.11.23.18.58.45.760/TruffleHotSpotCompilation-13746[Object#foo].bgv' 4
137 nodes, loops, branches

chrisseaton@Shopifys-MacBook-Pro truffleruby % jt graph --describe calls.rb 
Using TruffleRuby with Graal: mxbuild/truffleruby-jvm-ce
Dumping IGV graphs in /Users/chrisseaton/Documents/graalvm/truffleruby/graal_dumps/2021.11.23.19.00.01.002
[engine] opt done   id=1392  Object#foo                                         |Tier 2|Time   355( 184+171 )ms|AST   15|Inlined   1Y   0N|IR     84/   116|CodeSize     422|Timestamp 23041912391810|Src calls.rb:1
[engine] opt deopt  id=1393  GC.start                                                                                                                                       |Timestamp 23042014437358|Src gc.rb:86
[engine] opt deopt  id=1392  Object#foo                                                                                                                                     |Timestamp 23042016462449|Src calls.rb:1
$ /Users/chrisseaton/Documents/graalvm/truffleruby/tool/describe-graph 'graal_dumps/2021.11.23.19.00.01.002/TruffleHotSpotCompilation-13779[Object#foo].bgv' 4
84 nodes, calls, linear
```

eregon commented:
I think that would make more sense to have as an option or another subcommand of seafoam, e.g. `seafoam graph:n describe`.
That way, things like `org.graalvm.compiler.nodes.LoopBeginNode` are not hardcoded here but in Seafoam, where they need to be hardcoded anyway.

nirvdrum commented:
I ported this over to Seafoam where it's available as a new [`describe` sub-command](https://github.com/Shopify/seafoam/commit/918e19811263560e5edb7ae2b33fd30002e26a22) and updated the PR to use that sub-command.

chrisseaton commented:
I've added tests using this new functionality.

Benoit asked in the keyword arguments PR if the number of nodes expected should be fuzzy. I'm not sure - I've implemented the functionality but how about we wait and see if it's needed. An extra node or two here could be a legitimate regression.

chrisseaton commented:
Now I'm getting nervous is the number of nodes the same in Native...

nirvdrum commented:
I ran a quick example with:

```ruby
100_000.times { "abc".upcase }
```

Running:
`jt graph --describe --method 'String#upcase' test_file.rb`
`jt graph --describe --no-simplify --method 'String#upcase' test_file.rb`

Results:

| Build Type | Simplify Graphs? | Output                     |
|------------|------------------|----------------------------|
| JVM CE     | False            | 145 nodes, loops, branches |
| JVM CE     | True             | 203 nodes, loops, branches |
| Native     | False            | 216 nodes, loops, branches |
| Native     | True             | 216 nodes, loops, branches |

With graph simplification on by default, there's a pretty large node count difference between JVM CE and Native. It looks like graph simplification doesn't do anything with Native at all. I haven't investigated that yet.

eregon commented:
I doubt the number of nodes is really useful, it's just too variable and it might break when Graal decides to represent or canonicalize things slightly differently, and there is likely no time to investigate both graphs every time there is an extra or removed node. I think people would not be happy if small internal changes break those tests, and even less so to update such hardcoded numbers in tests.

As mentioned in https://github.com/Shopify/seafoam/pull/47/files#r768619417, I think it makes sense to count branches, loops and (non-inlined) calls.
Possibly also the number of reads and writes (I wouldn't differentiate between reading a field/array index).
Maybe even the number of guards, although that's likely also less stable.

The seafoam command could just provide counts for everything, and the regression tests would only use counts for the stable parts (e.g., not nb of `nodes`).

eregon commented on test/truffle/compiler/graphs/run:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env ruby
+
+NATIVE = `tool/jt.rb ruby --version` =~ /Native/
```
Use `TruffleRuby.native?` and remove the extra constant

chrisseaton replied:
Is TruffleRuby put on the path by JT?

eregon replied:
I see, this file is just run by `ruby` in PATH, that's an anti-pattern, it should use either truffleruby (via `jt ruby`) or `$SYSTEM_RUBY`.

eregon replied:
This is unused now

eregon commented on test/truffle/compiler/graphs/run:
```diff
@@ -0,0 +1,33 @@
+test 'test/truffle/compiler/graphs/args-req.rb', 36, 34, 'linear'
+test 'test/truffle/compiler/graphs/args-opt-set.rb', 36, 34, 'linear'
+test 'test/truffle/compiler/graphs/args-opt-unset.rb', 32, 31, 'linear'
+test 'test/truffle/compiler/graphs/args-rest.rb', 36, 34, 'linear'
```
I don't think we want to hardcode node numbers here, as explained in my other comment, because it's too brittle.
The fact it differs between native and not is already a bad sign for stability.

eregon commented on test/truffle/compiler/graphs/run:
```diff
@@ -0,0 +1,33 @@
+test 'test/truffle/compiler/graphs/args-opt-unset.rb', 32, 31, 'linear'
+test 'test/truffle/compiler/graphs/args-rest.rb', 36, 34, 'linear'
+
+puts "#$tests tests, #$failures failures"
```
```suggestion
puts "#{$tests tests}, #{$failures failures}"
```

eregon replied:
`#$` extrapolation seems very uncommon and particularly hard to read (also IDEs likely don't highlight it properly), I think we should avoid it, `#{...}` is much clearer.

chrisseaton replied:
`#{$tests tests}` isn't valid Ruby so I was confused. I see what you mean now.

eregon replied:
Still not fixed

eregon commented on tool/jt.rb:
```diff
@@ -794,6 +794,7 @@ def help
       jt profile                                     profiles an application, including the TruffleRuby runtime, and generates a flamegraph
       jt graph [ruby options] [--method Object#foo] [--watch] [--no-simplify] file.rb
                                                      render a graph of Object#foo within file.rb
+                              --describe             describe the shape of the graph (deopts, linear, branches, loops, calls)
```
`deopts` seems the least common, so I'd mention it last.

nirvdrum commented:
@eregon We've added a JSON output option to Seafoam. We haven't released the new gem with it yet, but I've pushed up a commit that changes the graph tests to use the JSON output to make the graph tests easier to work with. Let me know what you think. Key changes:

* Rather than compare strings, we can compare fields in a graph description `Hash`.
* Newly added features won't invalidate tests unless that feature is present in the graph and the test isn't explicitly looking for it.
* The node count can be supplied as a range to avoid slight changes in the node count from invalidating a test. If a number is supplied instead, it is converted to a range with only that number.

We can handle the node count a different way if you'd prefer. By comparing hashes, we could avoid the node count entirely. But, I think it'd be useful to see if the node count explodes.

eregon commented:
This looks much better and easier to evolve, it would be good to address the remaining comments (mostly small things).

I think an exact node count is only going to cause issues in CI and e.g. Graal/SVM people being angry they can't even change the internal graph representation to have a few more or less nodes without needing a PR in TruffleRuby.
Also the fact it differs between JVM and SVM is a clear sign it's not something sensible to hardcode.
It could be a Range, but it's also going to be arbitrary numbers and it would need to be pretty large like 0..100 or so, which makes it almost useless, but maybe better than nothing.

If you want to track the count or so, I think you'd need to count the number of reads, writes, If's, etc.
That should be quite a bit more stable.
If you want to compare the graph before & after in an automated manner then the graph should be serialized to some format, and we'd just store the result in the repo. It would need to be carefully designed to not break too often or easily with compiler or Truffle changes.

For the concrete case of argument handling shown here, I think it makes sense to track:
* stays linear (or 0 `if`, 0 `while`, 0 `unwind`)
* the number of reads, that should be fairly stable and something meaningful to optimize
* the number of writes (should be 0)
* 0 allocations
* 0 invokes

chrisseaton commented:
I've removed the number of nodes for now. The general shape (loops, linear) is enough now to write the tests we're depending on for keyword arguments. After merging it's then easy to iterate in the future to count calls, reads, and writes as Benoit suggests.

chrisseaton commented:
CF https://github.com/openjdk/jdk/pull/5135/files

eregon commented on spec/truffle/launcher_spec.rb:
```diff
@@ -80,7 +80,7 @@ def check_status_or_print(stdout_and_stderr)
     actual = Dir.children(File.dirname(RbConfig.ruby)).sort
     actual.delete('truffleruby.sh')
-    actual.should == known
+    actual.intersection(known).should == known
```
How can seafoam end up in truffleruby bin? Due to being gem installed?
I'd prefer we exclude it explicitly like `truffleruby.sh` then (which BTW can be removed, there is no more `truffleruby.sh`), otherwise this test tests nothing.

eregon replied:
> How can seafoam end up in truffleruby bin? Due to being gem installed?

It shouldn't end up there since seafoam should be installed on the ruby running `jt` and that should not be truffleruby.

eregon commented on tool/jt.rb:
```diff
@@ -2034,8 +2041,9 @@ def graph(*args)
-      sh env, 'gem', 'install', 'seafoam'
+    seafoam_version = '0.11'
+    unless Gem::Specification.find_all_by_name('seafoam').any? { |g| g.version == Gem::Version.new(seafoam_version) }
+      sh env, 'gem', 'install', 'seafoam', '-v', seafoam_version
```
This won't work in the internal CI (no internet access, should not install gems on the ruby in PATH).
Instead it should be done like `rubocop` with a `seafoam` helper method.
And then I'll add it to gem_test_pack.

eregon replied:
I can do this part
